temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityWorker.java

```diff
@@ -210,17 +210,20 @@ public void handle(ActivityTask task) throws Exception {
           if (info.hasDetails()) {
             canceledRequest.setDetails(info.getDetails());
           }
-          sendReply(
-              r,
-              new Result(r.getActivityId(), null, null, canceledRequest.build(), null),
-              metricsScope);
+          response =
+              new Result(r.getActivityId(), null, null, canceledRequest.build(), null, false);
+          sendReply(r, response, metricsScope);
         }
       } finally {
         MDC.remove(LoggerTag.ACTIVITY_ID);
         MDC.remove(LoggerTag.ACTIVITY_TYPE);
         MDC.remove(LoggerTag.WORKFLOW_ID);
         MDC.remove(LoggerTag.RUN_ID);
-        task.getCompletionHandle().apply();
+        // Apply completion handle if task has been completed synchronously or is async and manual
+        // completion hasn't been requested.
+        if (response != null && !response.isManualCompletion()) {
```

Comment:
Does response != null check mean that if activity throws an exception the counter is not decremented?

Reply:
No, it should be null only if activity is async.